analyzer/codechecker_analyzer/analyzers/cppcheck/analyzer.py

```diff
@@ -0,0 +1,350 @@
+
+from .. import analyzer_base
+
+from .config_handler import CppcheckConfigHandler
+from .result_handler import CppcheckResultHandler
+
+from ..config_handler import CheckerState
+
+LOG = get_logger('analyzer.cppcheck')
+
+
+def parse_checkers(cppcheck_output):
+    """
+    Parse cppcheck checkers list.
+    """
+    checkers = []
+
+    tree = ET.ElementTree(ET.fromstring(cppcheck_output))
+    root = tree.getroot()
+    errors = root.find('errors')
+    for error in errors.findall('error'):
+        name = error.attrib.get('id')
+        if name:
+            name = "cppcheck-" + name
+        msg = error.attrib.get('msg')
+        # TODO: Check severity handling in cppcheck
+        # severity = error.attrib.get('severity')
+
+        # checkers.append((name, msg, severity))
+        checkers.append((name, msg))
+
+    return checkers
+
+
+def parse_version(cppcheck_output):
+    """
+    Parse cppcheck version output and return the version number.
+    """
+    version_re = re.compile(r'^Cppcheck (?P<version>[\d\.]+)')
+    match = version_re.match(cppcheck_output)
+    if match:
+        return StrictVersion(match.group('version'))
+
+    return None
+
+
+class Cppcheck(analyzer_base.SourceAnalyzer):
+    """
+    Constructs the clang tidy analyzer commands.
+    """
+
+    ANALYZER_NAME = 'cppcheck'
+
+    def add_checker_config(self, checker_cfg):
+        LOG.error("Not implemented yet")
+
+    def get_analyzer_mentioned_files(self, output):
+        """
+        Return a collection of files that were mentioned by the analyzer in
+        its standard outputs, which should be analyzer_stdout or
+        analyzer_stderr from a result handler.
+        """
+        pass
+
+    def construct_analyzer_cmd(self, result_handler):
+        """
+        Construct analyzer command for cppcheck.
+        """
+        try:
+            config = self.config_handler
+
+            analyzer_cmd = [config.analyzer_binary]
+
+            # Enable or disable checkers.
+            enabled_severity_levels = set()
+
+            # Cppcheck runs with all checkers enabled for the time being
+            # the unneded checkers are passed as suppressed checkers
+            analyzer_cmd.append('--enable=all')
+
+            if enabled_severity_levels:
+                analyzer_cmd.append('--enable=' +
+                                    ','.join(enabled_severity_levels))
+
+            for checker_name, value in config.checks().items():
+                if value[0] == CheckerState.disabled:
+                    if checker_name.startswith("cppcheck-"):
+                        checker_name = checker_name[9:]
+                    # TODO python3.9 removeprefix method is better than lstrip
+                    analyzer_cmd.append('--suppress=' + checker_name)
+
+            # unusedFunction check is for whole program analysis,
+            # which is not compatible with per source file analysis.
+            analyzer_cmd.append('--suppress=unusedFunction')
```

Comment:
Could you please elaborate on what do you mean by the "configuration to support a 'default suppressed' "?

Reply:
Here we are hardcoding the fact that this one check should be disabled (by default? or always?) which in general mixes business logic and configuration that's... against high-level design conventions.

Currently, at least with CSA/CT the logic is that checks are grouped into profiles (with "default" being a special case!), and only checks that are in the enabled profiles (or explicitly enabled by the user) are run.

Do we want to explicitly forbid the user from ever running this check, for whatever reason?

The current configuration schema does not support this notion, so the idea is that we implement this notion (forbidden checks) into the configuration schema itself, and then we add this information into the cppcheck-specific config file, instead of hardcoding an identifier into the "driver" logic.

Reply:
@vodorok Perhaps this is is worth a higher-level follow-up feature request?

Reply:
True, it is in the making.